Properly handle errors in the cookie's domain when setting a cookie with the Cookie Store API

RupinMittal opened the pull request:
#### 58f6a2dd0bf84e0bba186126c6474a0ff1064353
<pre>
Properly handle errors in the cookie&apos;s domain when setting a cookie with the Cookie Store API
<a href="https://bugs.webkit.org/show_bug.cgi?id=259529">https://bugs.webkit.org/show_bug.cgi?id=259529</a>

Reviewed by Alex Christensen.

The spec (<a href="https://wicg.github.io/cookie-store/#set-cookie-algorithm)">https://wicg.github.io/cookie-store/#set-cookie-algorithm)</a>
dictates that in the set function, if the domain is not null, then if
the domain begins with a &apos;.&apos;, the promise should be rejected with a
TypeError. If the domain does not begin with a &apos;.&apos; but the url&apos;s host
is not equal to domain and it does not end with a &apos;.&apos; followed by the
domain, the promise should be rejected with a TypeError. Additionally,
if the byte sequence length of the domain (in UTF8 format) is greater
than the maximum attribute value size (current 1024 bytes according to
<a href="https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size)">https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size)</a>,
then the promise should be rejected with a TypeError. This patch adds
these checks.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_opaque_origin.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):

Canonical link: <a href="https://commits.webkit.org/266351@main">https://commits.webkit.org/266351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24fff31431b8915f695a3fd246770e7025861d1f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13590 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13904 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14237 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15326 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12909 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13671 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16412 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13984 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15601 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13756 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14397 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11498 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16025 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11685 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12253 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19296 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12760 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12421 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15633 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12941 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10834 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12211 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3313 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16541 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12784 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_opaque_origin.https-expected.txt:
```diff
@@ -1,4 +1,4 @@
 
-PASS cookieStore in non-sandboxed iframe should not throw
+FAIL cookieStore in non-sandboxed iframe should not throw assert_equals: cookieStore ${apiCall} should not throw expected "no exception" but got "TypeError"
```
I'm not entirely sure why this change happened yet. I'm looking into it, but still wanted to post the patch for review in the meantime.

achristensen07 replied:
This is fine.  Seems like a test just getting farther than it used to.

achristensen07 commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -211,9 +211,22 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+            return;
+        }
+        auto host = url.host();
+        if (host != cookie.domain && !host.endsWith(StringView { makeString(".", cookie.domain) })) {
```
We can probably do this without allocating another string with a dot prepended.

I don't see handling for this:
If the byte sequence length of encodedDomain is greater than the maximum attribute value size, then return failure.

Also, do we want to check the public suffix list?  If domain is "com" and url.host() is "apple.com" we probably also want to reject the promise.  This may be more of a spec issue though.

RupinMittal replied:
Good point. I've changed the check to work without allocating a new string and have added handling for the case where the byte sequence length of the domain is greater than the maximum attribute value size. 

Regarding the public suffix list--what you're saying makes sense and maybe we should check this. I didn't add it in the newest iteration because it's not in the spec, but perhaps this is something we should discuss.

achristensen07 commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -211,9 +211,31 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+
+        auto host = url.host();
+        if (host != cookie.domain) {
+            if (host.length() <= cookie.domain.length() || host.characterAt(host.length() - cookie.domain.length() - 1) != '.' || !host.endsWith(cookie.domain)) {
```
I'd write this with something like this:

if (!host.endsWith(cookie.domain) || !StringView(host).substring(0, host.length() - cookie.domain.length()).endsWith('.'))

RupinMittal replied:
I have two questions about this:

1. Doesn't `StringView::substring()` allocate a new `StringView`? Which is what we were trying to avoid?
2. If `host == cookie.domain`, then this if condition will be true because `!host.endsWith(cookie.domain)` will be false and `!StringView(host).substring(0, host.length() - cookie.domain.length()).endsWith('.')` will be true. So we will end up rejecting when we shouldn't correct?

If `StringView::substring()` does not allocate a new `StringView`, then maybe we can use something like this: `if (!host.endsWith(cookie.domain) || (host.length() > cookie.domain.length() && !StringView(host).substring(0, host.length() - cookie.domain.length()).endsWith('.')))`. Otherwise, maybe there is some other more concise way?

achristensen07 replied:
A StringView doesn't allocate anything.  It's just a few integers in registers.  makeString does allocate and copy.

You're right, you'll need some logic for if the strings are equal.  At the end of the day we're just talking about clean ways of writing the same thing.  If there's no way it can go out of bounds, they're all equivalent.

RupinMittal replied:
Ok that sounds good, I've updated the code to use `StringView::substring()` with the extra check as I wrote above.

achristensen07 commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -211,9 +211,31 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+            }
+        }
+
+        static constexpr auto maximumAttributeValueSize = 1024;
```
Could you add a spec link in a comment for this?  Otherwise it looks like just a magic number.

RupinMittal committed:
Properly handle errors in the cookie's domain when setting a cookie with the Cookie Store API
https://bugs.webkit.org/show_bug.cgi?id=259529

Reviewed by Alex Christensen.

The spec (https://wicg.github.io/cookie-store/#set-cookie-algorithm)
dictates that in the set function, if the domain is not null, then if
the domain begins with a '.', the promise should be rejected with a
TypeError. If the domain does not begin with a '.' but the url's host
is not equal to domain and it does not end with a '.' followed by the
domain, the promise should be rejected with a TypeError. Additionally,
if the byte sequence length of the domain (in UTF8 format) is greater
than the maximum attribute value size (current 1024 bytes according to
https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size),
then the promise should be rejected with a TypeError. This patch adds
these checks.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_opaque_origin.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):

Canonical link: https://commits.webkit.org/266351@main
